### ISSUE TITLE
extra 'load'

### ISSUE THREAD

**Issue by bpangamiasyscom** [llvm/llvm-project#58653]
Clang is creating two 'lw', but one 'lw' should be enough as demonsted by GCC:

https://godbolt.org/z/vG3qbKqPq

**Comment by topperc** [comment]
This issue doesn't appear to be unique to RISC-V. It happens on X86 for the same code. I think it's a middle end alias analysis issue.

**Comment by gandhi56** [comment]
Source code:
```
struct d{
    unsigned long long a:1;
    unsigned long long c;
    unsigned long long b:4;
};

struct d *ptr;
void foo(){
   ptr->a = 1;
   ptr->b = 0x8;
}
```

clang 10.0.1 x86-64 asm:
```
# clang -O2 -S -o - 
foo:                                    # @foo
        mov     rax, qword ptr [rip + ptr]
        or      byte ptr [rax], 1
        mov     rax, qword ptr [rip + ptr]
        mov     cl, byte ptr [rax + 16]
        and     cl, -16
        or      cl, 8
        mov     byte ptr [rax + 16], cl
        ret
```
clang trunk x86-64 asm:
```
# clang -O2 -S -o -
foo:                                    # @foo
        movq    ptr(%rip), %rax
        orb     $1, (%rax)
        movq    ptr(%rip), %rax
        movzbl  16(%rax), %ecx
        andb    $-16, %cl
        orb     $8, %cl
        movb    %cl, 16(%rax)
        retq
ptr:
        .quad   0
```

**Comment by topperc** [comment]
> Source code:
> 
> ```
> struct d{
>     unsigned long long a:1;
>     unsigned long long c;
>     unsigned long long b:4;
> };
> 
> struct d *ptr;
> void foo(){
>    ptr->a = 1;
>    ptr->b = 0x8;
> }
> ```
> 
> clang 10.0.1 x86-64 asm:
> 
> ```
> # clang -O2 -S -o - 
> foo:                                    # @foo
>         mov     rax, qword ptr [rip + ptr]
>         or      byte ptr [rax], 1
>         mov     rax, qword ptr [rip + ptr]
>         mov     cl, byte ptr [rax + 16]
>         and     cl, -16
>         or      cl, 8
>         mov     byte ptr [rax + 16], cl
>         ret
> ```
> 
> clang trunk x86-64 asm:
> 
> ```
> # clang -O2 -S -o -
> foo:                                    # @foo
>         movq    ptr(%rip), %rax
>         orb     $1, (%rax)
>         movq    ptr(%rip), %rax
>         movzbl  16(%rax), %ecx
>         andb    $-16, %cl
>         orb     $8, %cl
>         movb    %cl, 16(%rax)
>         retq
> ptr:
>         .quad   0
> ```

I believe the only differences between llvm 10 and trunk are due to trunk defaulting to -fpie.

**Comment by bpangamiasyscom** [comment]
@topperc I tried to use "-fno-pie", but there is no difference in generated code for riscv.

**Comment by topperc** [comment]
> @topperc I tried to use "-fno-pie", but there is no difference in generated code for riscv.

I wasn't trying to say it would. I was commenting on x86, but I think I got distracted by the intel and at&t syntax difference. I don't think -fpie made a difference after all.

The second "mov     rax, qword ptr [rip + ptr]" is the same bug for x86-64.

**Comment by BaoshanPang** [comment]
This is a long standing issue when bit field is invlolved,  it can be reproduced with clang 9.0(didn't try any older version).

**Comment by BaoshanPang** [comment]
> This issue doesn't appear to be unique to RISC-V. It happens on X86 for the same code. I think it's a middle end alias analysis issue.

Yes, it seems AA would think these two could be aliased(I changed variable name 'ptr' to 'xptr' to avoid confusion with the keyword 'ptr' used by LLVM):

%0 = load ptr, ptr @xptr, align 4, !tbaa !4
@xptr = dso_local local_unnamed_addr global ptr null, align 4
